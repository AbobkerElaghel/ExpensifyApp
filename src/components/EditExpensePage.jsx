import React from "react";
import ExpenseForm from "./ExpenseForm.jsx";
import {connect} from "react-redux";
import {editExpense} from "../actions/expensesAction";

const EditExpensePage = props => {
    return(
        <div>
            <h3>Edit Expense</h3>
            <ExpenseForm expense={props.expense}
                         onSumbit={expense => {
                            props.dispatch(editExpense(props.match.params.id, expense));
                            props.history.push("/");
            }}/>
        </div>
    );
};

const mapStateToProps = (state, props) => ({expense: state.expenses.find(expense => expense.id === props.match.params.id)});

export default connect(mapStateToProps)(EditExpensePage);