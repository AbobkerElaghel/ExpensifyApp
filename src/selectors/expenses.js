import moment from "moment";

// get visible expenses
export default (expenses, {text, sortBy, startDate, endDate}) => (expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
}).sort((a, b) => {
    if(sortBy === "amount"){
        return b.amount - a.amount;
    } else if (sortBy === "date") {
        return b.createdAt - a.createdAt;
    }}));
