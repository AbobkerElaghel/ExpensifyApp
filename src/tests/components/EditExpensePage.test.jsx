import React from "react";
import { shallow } from "enzyme";
import {EditExpensePage} from "../../components/EditExpensePage.jsx";
import expenses from "../fixtures/expenses";

let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;
beforeEach(() => {
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage expense={expenses[0]} editExpense={editExpenseSpy} removeExpense={removeExpenseSpy} history={historySpy}/>);
});

test('should render Expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
});

test('should handle remove expense', () => {
    wrapper.find("button").simulate('click');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
});