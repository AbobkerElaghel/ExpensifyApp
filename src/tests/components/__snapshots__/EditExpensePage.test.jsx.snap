// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense page correctly 1`] = `
<div>
  <h3>
    Edit Expense
  </h3>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "TESTING",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
